geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseIndexAdapter.scala

```diff
@@ -301,110 +303,131 @@ class HBaseIndexAdapter(ds: HBaseDataStore) extends IndexAdapter[HBaseDataStore]
           // this is unnecessary as we have already sorted and merged
-          val mrrf = new MultiRowRangeFilter(group)
           // note: mrrf first priority
-          s.setFilter(if (coprocessor || filters.nonEmpty) { new FilterList(filters.+:(mrrf): _*) } else { mrrf })
-        } else {
-          filters match {
-            case Nil    => // no-op
-            case Seq(f) => s.setFilter(f)
-            case f      => s.setFilter(new FilterList(f: _*))
-          }
+          filters.+:(new MultiRowRangeFilter(group))
         }
-
-        s.setCacheBlocks(cacheBlocks)
-        cacheSize.foreach(s.setCaching)
+        // note: coprocessors always expect a filter list
+        val filter = mrrf match {
+          case f if coprocessor || f.lengthCompare(1) > 0 => new FilterList(f: _*)
+          case f => f.headOption.orNull
+        }
+        scan.setFilter(filter)
+        scan.addFamily(colFamily).setCacheBlocks(cacheBlocks)
+        cacheSize.foreach(scan.setCaching)
 
         // apply visibilities
-        ds.applySecurity(s)
+        ds.applySecurity(scan)
 
-        groupedScans.add(s)
+        scan
       }
 
-      // TODO GEOMESA-1806 align partitions with region boundaries
-
-      if (!rowRanges.isEmpty) {
-        var i = 1
-        var groupStart = 0
-        var groupCount = 1
-        var groupFirstByte: Byte =
-          if (rowRanges.get(0).getStartRow.isEmpty) { 0 } else { rowRanges.get(0).getStartRow()(0) }
-
-        while (i < rowRanges.size()) {
-          val nextRange = rowRanges.get(i)
-          // add the group if we hit our group size or if we transition the first byte (i.e. our shard byte)
-          if (groupCount == rangesPerThread ||
-              (nextRange.getStartRow.length > 0 && groupFirstByte != nextRange.getStartRow()(0))) {
-            // note: excludes current range we're checking
-            addGroup(rowRanges.subList(groupStart, i))
-            groupFirstByte = if (nextRange.getStopRow.isEmpty) { Byte.MaxValue } else { nextRange.getStopRow()(0) }
-            groupStart = i
-            groupCount = 1
-          } else {
-            groupCount += 1
-          }
-          i += 1
+      val maxRangesPerGroup = {
+        def calcMax(maxPerGroup: Int, threads: Int): Int = {
+          val totalRanges = rangesPerRegionServer.values.map(_.length).sum
+          math.min(maxPerGroup, math.max(1, math.ceil(totalRanges.toDouble / threads).toInt))
         }
+        if (coprocessor) {
+          calcMax(ds.config.maxRangesPerCoprocessorScan, ds.config.coprocessorThreads)
+        } else {
+          calcMax(ds.config.maxRangesPerExtendedScan, ds.config.queryThreads)
+        }
+      }
+
+      val groupedScans = Seq.newBuilder[Scan]
 
-        // add the final group - there will always be at least one remaining range
-        addGroup(rowRanges.subList(groupStart, i))
+      rangesPerRegionServer.foreach { case (_, ranges) =>
```

Comment:
IntelliJ points out that 'ranges' here is already used above as a variable name.  Consider renaming on this line and the one below to clarify.

Reply:
wilco

Reply:
fixed